### ISSUE TITLE
Service starts too early to find resolutionKMS in Debian Testing

### ISSUE THREAD

**Issue by meeas** [vmware/open-vm-tools#214]
I'm running Debian Testing and just updated my install and found open-vm-tools failed to resize resolution in my VM after reboot.  If I started the service with systemctl by hand after I had logged in, everything worked fine.  Digging in the the logs, I found that when the service started on boot, it would throw this error in /var/log/vmware-vmsvc.log:

[Dec 09 23:18:53.585] [ warning] [resolutionCommon] resolutionCheckForKMS: No system support for resolutionKMS.

However when starting it by hand, /var/log/vmware-vmsvc.log would show:

[Dec 09 23:28:49.781] [ message] [resolutionCommon] resolutionCheckForKMS: System support available for resolutionKMS.
[Dec 09 23:28:49.781] [ message] [vmtoolsd] Plugin 'resolutionKMS' initialized.

I'm guessing that open-vm-tools is starting too early so I added a dependency to start after gdm.service which worked like a charm.  Not sure if this is the right permanent solution, but thought I'd pass it along.

$ cat /etc/systemd/system/multi-user.target.wants/open-vm-tools.service 
[Unit]
Description=Service for virtual machines hosted on VMware
Documentation=http://open-vm-tools.sourceforge.net/about.php
ConditionVirtualization=vmware
DefaultDependencies=no
After=gdm.service
Before=cloud-init-local.service

[Service]
ExecStart=/usr/bin/vmtoolsd
TimeoutStopSec=5
PrivateTmp=yes

[Install]
WantedBy=multi-user.target

**Comment by bzed** [comment]
10.2.0 will be available in unstable in a few hours. 10.1.15 is in testing.

@meeas could you give 10.2.0 a try and if the bug is still there, please let me know.
(also, reporting such bugs in the debian bugtracker would be the appropriate place!)

**Comment by jrespeto** [comment]
I saw this issue and tried this fix also to start vmtoolsd `After=display-manager.service`.
Looks to have resolved my issue. 

Also tested this on a fresh install of kali linux. 
```
vmtoolsd -v
VMware Tools daemon, version 10.3.0.5330 (build-8931395)

uname -a 
Linux Kali 4.18.0-kali2-amd64 #1 SMP Debian 4.18.10-2kali1 (2018-10-09) x86_64 GNU/Linux
```

```
# cat /etc/systemd/system/multi-user.target.wants/open-vm-tools.service 
[Unit]
Description=Service for virtual machines hosted on VMware
Documentation=http://open-vm-tools.sourceforge.net/about.php
ConditionVirtualization=vmware
DefaultDependencies=no
Before=cloud-init-local.service
RequiresMountsFor=/tmp
After=display-manager.service

[Service]
ExecStart=/usr/bin/vmtoolsd
TimeoutStopSec=5

[Install]
WantedBy=multi-user.target

```

**Comment by rhertzog** [comment]
This issue seems to still exist in version 10.3.5-1 that is in Debian Testing and Kali Linux. I'm not sure that delaying the start is the best choice but there must definitely be something to reload/restart vmtoolsd once the graphical session has been started. How is it supposed to work currently ? (Last bug report on the topic in Kali https://bugs.kali.org/view.php?id=5051)

**Comment by rhertzog** [comment]
I can reproduce this problem but not on every boot. I think it's a race condition between the load of drm/vmwgfx kernel modules and the start of the service. When the issue is present, I can see (in the logs, configured at debug level) some lines with "Failed to open DRM render node." and "Failed to open DRM card node.". And effectively a simple restart is enough to bring back things in correct order.

My proposed fix is thus to modify the .service file by adding an ExecStartPre command loading the appropriate kernel driver:
```
[Service]
ExecStartPre=-/sbin/modprobe vmwgfx
ExecStart=/usr/bin/vmtoolsd
TimeoutStopSec=5
```

With this, I can no longer reproduce the problem and we are also not delaying it further in the systemd dependency tree. The "After: display-manager.service" directive was IMO not a good idea, it goes contrary to the logic of "Before=cloud-init-local.service" and "DefaultDependencies=no" which tries to execute this service quite early in the boot process (while display-manager is not designed to run so early).

**Comment by rhertzog** [comment]
@oliverkurth Can you reassess this bug in the light of the information that I provided? I don't know what patch you were referring to but the missing patch is unlikely to be the problem unless the patch is not in the last released version that I tested (10.3.5).

What do you think of my suggested fix? The only downside is possibly that you might not want to try to load that driver on a headless VM. The debian packaging even splits the graphical part (which includes the resolutionKMS plugin) in a separate package. Thus I suggested in https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=915031#5 to add the "ExecStartPre" in a separate configuration file in the service "drop-in" directory (eg `/lib/systemd/system/open-vm-tools.service.d/desktop.conf`).

**Comment by mykmelez** [comment]
See https://communities.vmware.com/message/2816924 for additional reports of this issue in Ubuntu 18.10 and 16.04 LTS guests.

**Comment by bzed** [comment]
@oliverkurth what do you think about @rhertzog 's suggested fix?

**Comment by oliverkurth** [comment]
@thomashvmw can you have a look?

**Comment by bzed** [comment]
If the ExecStartPre=-/sbin/modprobe vmwgfx is the proper fix, then I'd suggest to add it as systemd override in the open-vm-tools-desktop package. This would ensure that it does not affact the server installation but also works for the desktop users.

**Comment by thomashvmw** [comment]
My preference would NOT be to force modprobe vmwgfx, since one may want to start up the system with vmwgfx disabled, and it might be pretty hard to track down what's actually loading vmwgfx. As a developer I frequently boot the system without vmwgfx to avoid bugs created during development.

We had a previous issue on this and there IIRC the conclusion was that the best would be to make sure the open-vm-tools service was launched *after* the service that normally loads the vmwgfx module. ("graphical?"). Let me see if I can dig up that issue.

**Comment by thomashvmw** [comment]
So it's actually Issue #253.
The suggestion is to add
After=graphical.target
Under the Unit section of
/etc/systemd/system/multi-user.target.wants/open-vm-tools.service

Courtesy of @pha5matis

**Comment by bzed** [comment]
No,  that is way too late as it needs to be available for cloud-init before the network comes up.

**Comment by thomashvmw** [comment]
OK, so is there a way to honor the modprobe.blacklist=vmwgfx boot option with your solution or will it help to add a blacklist option to vmwgfx itself?

**Comment by rhertzog** [comment]
On Wed, 13 Mar 2019, thomashvmw wrote:
> OK, so is there a way to honor the modprobe.blacklist=vmwgfx boot option
> with your solution or will it help to add a blacklist option to vmwgfx
> itself?

Maybe we can use "modprobe --use-blacklist vmwgfx" in the systemd unit?

It might need some testing to ensure that it honors the kernel command
line as well and that it does what you want.

Cheers,
-- 
Raphaël Hertzog ◈ Debian Developer

Support Debian LTS: https://www.freexian.com/services/debian-lts.html
Learn to master Debian: https://debian-handbook.info/get/

**Comment by bzed** [comment]
Could somebody with a desktop in VMware please give it a try?

**Comment by thomashvmw** [comment]
I can confirm that "modprobe -b vmwgfx" or "modprobe --use-blacklist vmwgfx" honors the "modprobe.blacklist=vmwgfx" command line option when quickly tested on ubuntu and fedora so that should work fine.

**Comment by bzed** [comment]
@thomashvmw @rhertzog after reading https://www.freedesktop.org/software/systemd/man/modules-load.d.html# I think the better option would be to let systemd-modules-load.service handle this. A quick look on the systemd source gave me the impression, that this will handle the blacklist just fine.

We actually have a file there which load the fuse module, so I'll add vmwgfx.

**Comment by thomashvmw** [comment]
Sounds fine with me!

**Comment by huornlmj** [comment]
This issue is still present in Ubuntu 18.04 on VMware work station 16.1.1 build-17801498.
